Changes to when_floating rules

karthink opened the pull request:
when_floating is None by default
when set to True, apply rule to floating windows only
when set to False, apply rule to tiled windows only

As per discussion in #2846.

I haven't been able to test this code. When I start Qtile with startx with the path to this directory, my Xserver crashes. Is there an easier way to test Qtile?

m-col commented:
Awesome, thanks!

>  Is there an easier way to test Qtile?

In the source directory, check out `scripts/xephyr`. It uses `Xephyr` to run a nested X session in a window within your main X session. For this you will need `Xephyr` installed. Check out http://docs.qtile.org/en/latest/manual/hacking.html#using-xephyr, and dependencies are also listed at the top of that page.

m-col commented:
Are there any docs that need updating to reflect this?

Also it might be a good idea to add an entry to CHANGELOG about this so others can find the feature!

m-col commented:
Looks like your branch has conflicts with master so you'll need to rebase. If needed, there is some information on how to do this at: http://docs.qtile.org/en/latest/manual/howto/git.html#i-was-told-to-rebase-my-work

It might be good to also squash your commits down into a single commit while you're there - there's also a section on that page for that in case you find it useful.

karthink commented:
I added a line in the example code illustrating the difference.

m-col commented:
The changes look good but the CI is complaining about inconsistent indentation in the changelog entry. Make sure that the indentation of your entry matches the rest of the file.

elParaguayo commented:
Can we fix the indentation in changelog, resolve the conflict and rebase into a single commit so we can get this merged?

karthink commented:
I fixed the indentation and rebased on the latest master.

elParaguayo ***@***.***> writes:

> Can we fix the indentation in changelog, resolve the conflict and rebase into a single commit so we can get this merged?
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/qtile/qtile/pull/2900#issuecomment-986193422

elParaguayo commented:
Can you squash everything into a single commit (see http://docs.qtile.org/en/latest/manual/howto/git.html#i-was-told-to-squash-some-commits).

karthink commented:
OK, I think I managed to rebase on master and squash commits. Hope it looks right now.

elParaguayo commented on libqtile/lazy.py:
```diff
@@ -47,7 +47,7 @@ def __init__(self, call: CommandGraphCall, args: Tuple, kwargs: Dict) -> None:
 
         self._layouts: Set[str] = set()
-        self._when_floating = True
+        self._when_floating = None
```
```suggestion
        self._when_floating: Optional[bool] = None
```
This is the initial declaration of `when_floating` so you need to declare the type here.

This should fix the mypy error you're seeing.
